### PR TITLE
add descriptor to faker generator. add to_json method for serialization

### PR DESCRIPTION
**Pull request by matthiasdv**
I've added a few functions to the random_generators.py file fore the purpose of serializing generators to JSON-format. The idea was that the `description()` method provides a dict buffer that can be serialised by a json_dumps call. 

I used the merge_2_dicts util function, but maybe in a bit of a clumsy way. 

Feel free to comment and review. 

I wasn't able to set up my pipenv environment due to the following issue: 

[pip install bson error, pip verion is 10.0.0](https://github.com/py-bson/bson/issues/77)

In short; 'pipenv --three' uses v10 of pip, while this version breaks an import statement in the bson module. The most straightforward way of solving this is downgrading to pip v 9.x. Is there a way for pipenv to specify the pip version? 

tried calling the unit tests without the pipenv environment, but bumped into an unknown module `path`

Feedback is appreciated.

### CODE REVIEW

**Comment by sv3ndk** [comment]
Hi Matthias, 

Thanks a lot for spending your time and effort on Trumania, that's exciting! I'm happy to look at your review. 

Concerning pip and pipenv, I confirm that my current Trumania pipenv environment reports that's it's using pip 9, it was probably the latest version when I created it:  

```sh
> pipenv run pip freeze
attrs==17.4.0
bson==0.5.0
decorator==4.1.2
Faker==0.8.9
networkx==2.0
numpy==1.14.0
pandas==0.22.0
path.py==10.5
pluggy==0.6.0
py==1.5.2
pymongo==3.6.0
pytest==3.3.2
python-dateutil==2.6.1
pytz==2017.3
scipy==1.0.0
six==1.11.0
tabulate==0.8.2
text-unidecode==1.1
-e git+git@github.com:RealImpactAnalytics/trumania.git@2d7b6dc85eadf61828de2faf23c60c0026c63120#egg=trumania
You are using pip version 9.0.1, however version 10.0.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
```

After reading your message I don't dare touching it now :) 

Could you try if this command successfully downgrades pip on your side?

```sh
pipenv run pip install pip==9.0.1
```

I also see discussions along the same line in this issue: https://github.com/pypa/pipenv/issues/1996

**Comment by sv3ndk** [comment]
Concerning the PR content itself, it's a good idea to add a `description()` to the `FakerGenerator()`, thanks for looking into that. 

Also, nice usage of `to_json`, I should have thought of using that for either `description()` or the `save_to_db()` and `load_from()` methods. 


As a bit of background, in Trumania,`description()` is meant to produce a human-readable json while `save_to_db()` and `load_from()` handle json as saved to file. `description()` is therefore sometimes a bit lossy in case too much details hurt redeability (for example [Population.description()](https://github.com/RealImpactAnalytics/trumania/blob/master/trumania/core/population.py#L179 ) just provide the names of the relationships, without their content and stuff ). There's also a [Circus.description()](https://github.com/RealImpactAnalytics/trumania/blob/master/trumania/core/circus.py#L268) on the top-level circus that builds a tree of the full structure )

Do you confirm your intention is to display the generator parameter to users? If so, I'd rather keep the nested structure, something like below. I admit `merge_2_dict`would work as well, though I feel the flat structure it produces make the result a tad less readable.


```python 
{
       "type": "NumpyRandomGenerator",
       "method": self.method.__name__,
       "numpy_parameters": self.numpy_parameters
}
```


```python
{
       "type": "FakerGenerator",
       "method": self.method.__name__,
       "seed": self.seed,
       "faker_parameters": self.numpy_parameters
}
```

Now I admit that in many cases, the human-friendly representation of `description()` contains just as much info than in the technical content saved to file by `save_to_db()`, so maybe we should rationalise all that in a next step. 

What do you think?

**Comment by matthiasdv** [comment]
Thanks for the extensive comments. 

`pipenv run pip install pip==9.0.1` did not seem to fix the issue. the `--three` flag in the `pipenv install --three` command seems to forcibly update pip to 10.x again. I've managed to circumvent the problem by simply omitting the `--three` flag.

The pluggy dependency also threw a hash mismatch error during installation, but solved this by manually installing using `pipenv run pip install pluggy==0.6.0` 

after those two steps pipenv check succeeds, and I can successfully run all Trumania tests : )

regarding the PR itself; From the way you describe the `description()` method I get the feeling I'm misusing it. I simply want to serialize the parameters passed to the `Generator`'s generation method for later use. Seems like `save_to_db()` would be a better fit for the job. 

Didn't know about `Circus.description()` though. I wrote a component that replicates similar functionality.

I should probably refactor to use `save_to_db()` rather than `description()`. 

I'm currently writing a set of components that observe a given data source, and then  return a set of Trumania operations to replicate the observed set with fake data. Maybe I should integrate this a bit more closely with Trumania itself?
